// overwatch-api/lib/parser/stats.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (platform, region, tag, cb) {

  var url = platform === 'pc' ? 'https://playoverwatch.com/en-us/career/' + platform + '/' + region + '/' + tag : 'https://playoverwatch.com/en-us/career/' + platform + '/' + tag;

  var options = {
    uri: encodeURI(url),
    encoding: 'utf8'
  };

  (0, _requestPromise2.default)(options).then(function (htmlString) {

    // Begin html parsing.
    var $ = _cheerio2.default.load(htmlString);
    var user = $('.header-masthead').text();
    var level = $('.player-level div').first().text();
    var portrait = $('.player-portrait').attr('src');
    var permission = $('.masthead-permission-level-text').text();

    // Get prestige level by matching .player-level background url hex.
    var prestigeEl = $('.player-level').first().attr('style');
    var prestigeHex = prestigeEl.match(/0x0*[1-9a-fA-F][0-9a-fA-F]*/);
    var prestigeLevel = prestigeHex ? (0, _utils.getPrestigeLevel)(prestigeHex[0]) : 0;

    // Endorsements.
    var endorsementLevel = $('.masthead .endorsement-level div').last().text();
    var endorsementFrame = $('.masthead .EndorsementIcon').attr('style').slice(21, -1);

    var sportsmanshipValue = $('.masthead .EndorsementIcon-border--sportsmanship').data('value');
    var shotcallerValue = $('.masthead .EndorsementIcon-border--shotcaller').data('value');
    var teammateValue = $('.masthead .EndorsementIcon-border--teammate').data('value');

    var endorsement = {
      sportsmanship: { value: sportsmanshipValue, rate: parseFloat((sportsmanshipValue * 100).toFixed(2)) },
      shotcaller: { value: shotcallerValue, rate: parseFloat((shotcallerValue * 100).toFixed(2)) },
      teammate: { value: teammateValue, rate: parseFloat((teammateValue * 100).toFixed(2)) },
      level: parseInt(endorsementLevel),
      frame: endorsementFrame
    };

    endorsement.icon = (0, _svg.createEndorsementSVG)(endorsement);

    var stats = {};

    //
    // Top Heroes.
    //
    var topHeroCategories = {
      quickplay: {
        'played': '0x0860000000000021',
        'games_won': '0x0860000000000039'
      },
      competitive: {
        'played': '0x0860000000000021',
        'games_won': '0x0860000000000039',
        'win_rate': '0x08600000000003D1'
      }
    };

    // Quickplay.
    stats['top_heroes'] = { quickplay: {} };
    Object.keys(topHeroCategories.quickplay).forEach(function (k) {
      var topHeroesEls = $('#quickplay [data-category-id="overwatch.guid.' + topHeroCategories.quickplay[k] + '"]').find('.progress-category-item');
      var topHeroes = [];
      topHeroesEls.each(function (i, el) {
        var stat = {};
        stat.hero = $(this).find('.title').text();
        stat.img = $(this).find('img').attr('src');
        stat[k] = $(this).find('.description').text();
        topHeroes.push(stat);
      });
      stats['top_heroes']['quickplay'][k] = topHeroes;
    });

    // Competitive.
    stats['top_heroes']['competitive'] = {};
    Object.keys(topHeroCategories.competitive).forEach(function (k) {
      var topHeroesEls = $('#competitive [data-category-id="overwatch.guid.' + topHeroCategories.competitive[k] + '"]').find('.progress-category-item');
      var topHeroes = [];
      topHeroesEls.each(function (i, el) {
        var stat = {};
        stat.hero = $(this).find('.title').text();
        stat.img = $(this).find('img').attr('src');
        stat[k] = $(this).find('.description').text();
        topHeroes.push(stat);
      });
      stats['top_heroes']['competitive'][k] = topHeroes;
    });

    //
    // Career Stats
    //
    var statCategories = ['Combat', 'Match Awards', 'Assists', 'Average', 'Miscellaneous', 'Best', 'Game'];

    // Quickplay Stats.
    statCategories.forEach(function (item) {
      var els = $('#quickplay [data-category-id="0x02E00000FFFFFFFF"] h5:contains("' + item + '")').closest('table').find('tbody tr');
      var statsArr = [];
      els.each(function (i, el) {
        var stat = {};
        stat.title = $(this).find('td').first().text();
        stat.value = $(this).find('td').next().text();
        statsArr.push(stat);
      });
      item = item.replace(' ', '_').toLowerCase();
      stats[item] = { quickplay: [] };
      stats[item]['quickplay'] = statsArr;
    });

    // Competitive Stats.
    statCategories.forEach(function (item) {
      var els = $('#competitive [data-category-id="0x02E00000FFFFFFFF"] h5:contains("' + item + '")').closest('table').find('tbody tr');
      var statsArr = [];
      els.each(function (i, el) {
        var stat = {};
        stat.title = $(this).find('td').first().text();
        stat.value = $(this).find('td').next().text();
        statsArr.push(stat);
      });
      item = item.replace(' ', '_').toLowerCase();
      stats[item]['competitive'] = [];
      stats[item]['competitive'] = statsArr;
    });

    var json = {
      username: user,
      level: parseInt(level) + prestigeLevel,
      portrait: portrait,
      endorsement: endorsement,
      private: permission === 'Private Profile',
      stats: stats
    };

    cb(null, json);
  }).catch(function (err) {
    cb(err);
  });
};

var _cheerio = require('cheerio');

var _cheerio2 = _interopRequireDefault(_cheerio);

var _requestPromise = require('request-promise');

var _requestPromise2 = _interopRequireDefault(_requestPromise);

var _utils = require('./utils');

var _svg = require('./svg');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
